Healthy/Classes/Modules/Search/Search/SearchViewModel.swift

```diff
@@ -1,15 +1,108 @@
-// MARK: SearchViewModel
+import Combine
 
-final class SearchViewModel {}
+ protocol SearchDataSource {
+    func loadRecipes() async throws -> [Recipe]
+ }
 
-// MARK: SearchViewModel
+ // MARK: SearchViewModel
 
-extension SearchViewModel: SearchViewModelInput {}
+ final class SearchViewModel {
+    private var subscriptions = Set<AnyCancellable>()
+    @Published private(set) var searchKeyword = ""
+    @Published private(set) var searchFilter = SearchFilter()
+    @Published private(set) var recipes = [Recipe]()
+    @Published private var state = SearchState.initial
 
-// MARK: SearchViewModelOutput
+    private let searchDataSource: SearchDataSource
 
-extension SearchViewModel: SearchViewModelOutput {}
+    init(searchDataSource: SearchDataSource) {
+        self.searchDataSource = searchDataSource
+    }
+ }
 
-// MARK: Private Handlers
+ // MARK: SearchViewModel
 
-private extension SearchViewModel {}
+ extension SearchViewModel: SearchViewModelInput {
+    func addNewSearchKeyword(_ keyword: String) {
+        searchKeyword = keyword
+        updateSearch()
+    }
+
+    func addNewSearchFilter(_ filter: SearchFilter) {
+        searchFilter = filter
+        updateSearch()
+    }
+ }
+
+ // MARK: SearchViewModelOutput
+
+ extension SearchViewModel: SearchViewModelOutput {
+    var recipesPublisher: AnyPublisher<[Recipe], Never> {
+        $recipes.eraseToAnyPublisher()
+    }
+
+    var errorPublisher: AnyPublisher<Error, Never> {
+        $state
+            .compactMap {
+            guard case let .failure(error) = $0 else { return nil }
+                return error
+        }
+        .eraseToAnyPublisher()
+    }
+
+    var isEmptyPublisher: AnyPublisher<Bool, Never> {
+        $recipes
+            .map { $0.isEmpty }
+            .eraseToAnyPublisher()
+    }
+
+    var isLoadingPublisher: AnyPublisher<Bool, Never> {
+        $state
+            .map {
+                guard case .loading = $0 else { return false }
+                return true
+            }
+            .eraseToAnyPublisher()
+    }
+
+    var isLoadingMore: AnyPublisher<Bool, Never> {
+        $state
+            .map {
+                guard case .loadingMore = $0 else { return false }
+                return true
+            }
+            .eraseToAnyPublisher()
+    }
+
+    var isLoaded: AnyPublisher<Bool, Never> {
+        $state
+            .map {
+                guard case .loaded = $0 else { return false}
+                return true
+            }
+            .eraseToAnyPublisher()
+    }
+ }
+
+    // MARK: Search states
+
+    enum SearchState {
+        case initial, loading, loadingMore, loaded
+        case failure(Error)
+    }
```

Comment:
```suggestion
    // MARK: Nested Types

    enum SearchState {
        case initial, loading, loadingMore, loaded
        case failure(Error)
    }
```

Should be moved to the end and should be private.

عايشين في ملكوت الله - Please fix the indentation 😅